google-cloud-firestore/acceptance/firestore/query_test.rb

```diff
@@ -284,4 +284,37 @@
     _(results.map(&:document_id)).must_equal ["doc1", "doc2"]
     _(results.map { |doc| doc[:foo] }).must_equal ["a", "b"]
   end
+
+  it "has to_json method and from_json class method" do
+    rand_query_col = firestore.col "#{root_path}/query/#{SecureRandom.hex(4)}"
+    rand_query_col.doc("doc1").create({foo: "a"})
+    rand_query_col.doc("doc2").create({foo: "b"})
+    rand_query_col.doc("doc3").create({foo: "c"})
+
+    original_query = rand_query_col.order(:foo).limit_to_last 2
+
+    json = original_query.to_json
+    _(json).must_be_instance_of String
+
+    query = Google::Cloud::Firestore::Query.from_json json, firestore
```

Comment:
Wasn't it QueryPartition that we needed to serialize/deserialize?

Reply:
Sorry, I should have explained why I changed my mind about that. The `Query` returned by `QueryPartition#to_query` is the equivalent of the `QueryPartition` that produced it. (It is the ordered collection group query held by the `QueryPartition`, plus the `start_at` and `end_before` constraints held separately by the `QueryPartition`.) I thought I'd see if this an acceptable serialization before adding `to_json` and `from_json` to `QueryPartition`. The methods needed to be added to `Query` in any case.

Reply:
Ah okay, that makes sense. The `Query` serialization LGTM. If we just layer on a similar serialization for the `QueryPartition` fields, I think we're good to go.

Reply:
@crwilcox @schmidt-sebastian  Regarding this:

> It should be possible for a user to serialize the result of the QueryPartition API so that the partitions can be processed on other machines.

Should the `QueryPartition` be serializable or is it enough to serialize the `Query` returned by `#to_query`?

```rb
partitions = col_group.partitions 3

data = partitions[0].to_query.to_json

# Somewhere else
query = Google::Cloud::Firestore::Query.from_json data, client
```

Reply:
@dazuma I tried adding `QueryPartition#to_json` and `QueryPartition.from_json` today and the result was surprisingly messy. I can certainly add these methods if you want, but I would prefer to just rely on `Query#to_json` and `Query.from_json`.

Reply:
ok we can keep it as is.